sklearn/_fast_gradient_boosting/splitting.pyx

```diff
@@ -0,0 +1,737 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+"""This module contains routines and data structures to:
+
+- Find the best possible split of a node. For a given node, a split is
+  characterized by a feature and a bin.
+- Apply a split to a node, i.e. split the indices of the samples at the node
+  into the newly created left and right childs.
+"""
+cimport cython
+from cython.parallel import prange
+import numpy as np
+cimport numpy as np
+from openmp cimport omp_get_max_threads
+from libc.stdlib cimport malloc, free
+from libc.string cimport memcpy
+
+from .histogram cimport _build_histogram
+from .histogram cimport _build_histogram_no_hessian
+from .histogram cimport _build_histogram_root
+from .histogram cimport _build_histogram_root_no_hessian
+from .histogram cimport _subtract_histograms
+# from .histogram cimport _subtract_histograms
+from .types cimport X_BINNED_DTYPE_C
+from .types cimport Y_DTYPE_C
+from .types cimport G_H_DTYPE_C
+from .types cimport hist_struct
+from .types import HISTOGRAM_DTYPE
+
+
+# Note: in a lot of functions here, we pass feature_idx and the whole 2d
+# histograms arrays instead a lot just histograms[feature_idx]. This is
+# because Cython generated C code will have strange Python interactions (likely
+# related to the GIL release and the custom histogram dtype) when using 1d
+# histogram arrays.
+
+# epsilon for comparing gains to avoid floating precision issues that might be
+# caused by the (slightly non-deterministic) parallel sums over gradients and
+# hessians
+cdef Y_DTYPE_C EPS = 1e-13
```

Comment:
```suggestion
cdef Y_DTYPE_C EPS = np.finfo(np.float64).eps
```

Reply:
I don't think `np.finfo(np.float64).eps` is appropriate, the difference might be bigger than this. This is used for **gain comparison**. The difference between 2 gains that should be equal is caused by non-deterministic summing over the gradients and hessians array. But gradients are squared in the gain computation, and we divide by the hessians. So the difference is very likely to be more than `np.finfo(np.float64).eps`